core/sequencer/server.go

```diff
@@ -138,53 +137,55 @@ func (s *Server) RunBatch(ctx context.Context, in *spb.RunBatchRequest) (*empty.
 	// Count items to be processed.  Unfortunately, this means we will be
 	// reading the items to be processed twice.  Once, here in RunBatch and
 	// again in CreateEpoch.
-	metadata := &spb.MapMetadata{
-		Source: &spb.MapMetadata_SourceSlice{
-			LowestWatermark:  lastMeta.GetSource().GetHighestWatermark(),
+	sources := make(map[int64]*spb.MapMetadata_SourceSlice)
+	for sliceID, high := range highs {
+		sources[sliceID] = &spb.MapMetadata_SourceSlice{
+			LowestWatermark:  lastMeta.Sources[sliceID].GetHighestWatermark(),
 			HighestWatermark: high,
-		},
+		}
 	}
 
-	msgs, err := s.readMessages(ctx, in.DomainId, metadata.GetSource())
+	msgs, err := s.readMessages(ctx, in.DomainId, sources)
 	if err != nil {
 		return nil, err
 	}
-	if int32(len(msgs)) < in.MinBatch {
+	if len(msgs) < int(in.MinBatch) {
```

Comment:
The reason we want `MinBatch != 0` is to avoid creating a lot of empty batches. Each new batch creates additional data for clients to download -- and we're running this function every 100ms or so because we want new messages to be picked up quickly.

The plan is to replace `MinBatch` with a comparison of watermarks and timestamps.

Reply:
The TODO is #1047